### PR TITLE
Do async site scan on plugin activation instead of synchronous validation

### LINKED ISSUE #5101: Discontinue doing synchronous validation requests upon activating plugins
## Feature description

Depends on #4719.

Relates to #2069.

In #5100 no longer is a synchronous validation request performed when switching template modes. However, what remains is the synchronous validation request when activating plugins.

The validation errors are obtained in `AMP_Validation_Manager::validate_after_plugin_activation()` and then displayed via `AMP_Validation_Manager::print_plugin_notice()`.

Instead, upon activating a plugin when the “Plugin activated” notice is displayed:

image

There should at that point also be an admin notice from the AMP plugin which is accompanied by an enqueued script which does an _asynchronous_ request to check the site for validation issues. A spinner should be shown along with a message saying “checking for AMP compatibility issues”. Then once the results are back, a message can be shown saying there is no validation errors or that there are errors present with those activated plugins, and a link to see what they are. 

As it is currently, this admin notice should probably be suppressed if the user doesn't have developer tools enabled.

There could also be a good tie-in here with Plugin Suppression to allow a user to easily deactivate the plugins on AMP responses.

Relates to wizard compatibility scanning in #4795 and #4719.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

1. All existing server-side synchronous logic is removed.
1. Upon selecting to activate a plugin, the user is immediately taken to the resulting page with the normal "Plugin activated" notice shown.
1. On the resulting page, another "Plugin scanning" notice is displayed with a loading spinner.
1. REST API requests are initiated to do a site scan (#4719). 
1. The plugin sources for _unreviewed_ invalid markup are then aggregated from the REST API responses.
1. If there are plugin sources for unreviewed invalid markup:
   1. Give the notice a `warning` class.
   2. List out the plugins which are causing unreviewed validation errors.
   3. List out the validated URLs on which the errors occur. (The validated URLs on which a plugin's errors occur should probably be listed under the plugin itself.)
   4. Link to Plugin Suppression screen so user has ability to prevent plugins from running on AMP pages. (It's likely overkill to provide a way to suppress plugins directly from this notice.)
1. If there are no unreviewed validation errors, then the notice should be given a `success` class.

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by delawski**
## Summary

Fixes #5101

https://user-images.githubusercontent.com/478735/143105528-764a4fd6-73cc-43a4-ac46-7fb8eaee95fb.mp4

This PR renders an admin notice box whenever a plugin is activated. The admin notice starts a full Site Scan immediately and presents validation results as soon as the scan is complete:

| Scan in progress after plugin activation | Scan complete - no plugin issues |
| --- | --- |
| Screenshot 2021-11-23 at 21 09 12 | Screenshot 2021-11-23 at 21 12 33 |

| Scan complete - URLs list collapsed | Scan complete - URLs list expanded |
| --- | --- |
| Screenshot 2021-11-23 at 21 07 37 | Screenshot 2021-11-23 at 21 07 25 |
 
Note that there is a slight difference in the implementation compared to the design. After Site Scan is done, we're listing out all plugins causing AMP incompatibilities along with the URLs on which the issues were discovered. Also, a CTA to AMP compatible plugins is not hidden inside the `<details>` element but it's rather outside of it (since there might be multiple `<details>` elements in a single notice.

According to the design, we should be showing only issues related to the plugin (or multiple plugins) that has just been activated. The CTA should be collapsed by default inside the `<details>` element.

On one hand, the designed approach would be more complex in terms of the implementation. Since Core doesn't store a list of plugins that were last activated (or is it?), we would have to persist it on our own, probably using the Options API. On the other hand, though, from a user point of view, I believe it's better to let a user know about every plugin that causes AMP validation errors, not just the one that has been activated.

In any case, I'm happy to change the implementation so that it matches the design.

## Implementation Details
- A new JS entry point has been added to the `admin` group: `amp-site-scan-notice`.
- Since the admin notice requires the validation issues to be grouped by a different "dimension", i.e. not by a URL but rather a plugin, the `getSlugsFromValidationResults` helper had to be transformed into a more robust `getSourcesFromScannableUrls` function. The arrays returned by `getSlugsFromValidationResults` contained just the plugin/theme slugs like `[ 'amp', 'pwa' ]`. With `getSourcesFromScannableUrls`, the returned arrays also contain a list of URLs where the issues were found, e.g. `[ { slug: 'amp', urls: [ ... ] }, { slug: 'pwa', urls: [ ... ] } ]`.
- The PHP code related to the post-plugin-activation admin notice has been refactored out of `class-amp-validation-manager.php` and into a new class `PluginActivationSiteScan`.

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

**Comment by delawski** [comment]
Note that this PR, E2E tests in particular, will need to be updated once #6747 is merged into `develop`.

File: assets/src/admin/site-scan-notice/notice.js
Comment (dhaval-parekh) [inline assets/src/admin/site-scan-notice/notice.js:109]: Can we change the parent tag from `p` to something to `span`? because the `Loading` component uses the `div` tag which causes below error in the browser.

```
Warning: validateDOMNesting(...): <div> cannot appear as a descendant of <p>.
div
Loading@https://amp-dev.test/wp-content/plugins/amp/assets/js/amp-site-scan-notice.js?ver=f1872f78b895de49a60ee614caa3a22f:786:7
p
div
AdminNotice@https://amp-dev.test/wp-content/plugins/amp/assets/js/amp-site-scan-notice.js?ver=f1872f78b895de49a60ee614caa3a22f:367:7
SiteScanNotice@https://amp-dev.test/wp-content/plugins/amp/assets/js/amp-site-scan-notice.js?ver=f1872f78b895de49a60ee614caa3a22f:66:69
SiteScanContextProvider@https://amp-dev.test/wp-content/plugins/amp/assets/js/amp-site-scan-notice.js?ver=f1872f78b895de49a60ee614caa3a22f:1564:7
PluginsContextProvider@https://amp-dev.test/wp-content/plugins/amp/assets/js/amp-site-scan-notice.js?ver=f1872f78b895de49a60ee614caa3a22f:1154:7
UserContextProvider@https://amp-dev.test/wp-content/plugins/amp/assets/js/amp-site-scan-notice.js?ver=f1872f78b895de49a60ee614caa3a22f:1913:7
OptionsContextProvider@https://amp-dev.test/wp-content/plugins/amp/assets/js/amp-site-scan-notice.js?ver=f1872f78b895de49a60ee614caa3a22f:869:7
ErrorBoundary@https://amp-dev.test/wp-content/plugins/amp/assets/js/amp-site-scan-notice.js?ver=f1872f78b895de49a60ee614caa3a22f:530:5
ErrorContextProvider@https://amp-dev.test/wp-content/plugins/amp/assets/js/amp-site-scan-notice.js?ver=f1872f78b895de49a60ee614caa3a22f:623:7
Providers@https://amp-dev.test/wp-content/plugins/amp/assets/js/amp-site-scan-notice.js?ver=f1872f78b895de49a60ee614caa3a22f:3600:7 react-dom.860550b4.js:61:32
```

**Comment by dhaval-parekh** [comment]
Codewise PR looks good to me 👍

**Review by dhaval-parekh — APPROVED** [review]

File: assets/src/admin/site-scan-notice/notice.js
Comment (delawski) [reply assets/src/admin/site-scan-notice/notice.js:109]: That's a good catch!

Instead of changing the parent tag, the `<Loading inline={ true } />` component will now render as a `<span>` and not as a `<div>` (see 5253a32).

File: assets/src/components/admin-notice/style.scss
Comment (westonruter) [inline assets/src/components/admin-notice/style.scss]: The class names in this stylesheet should have an `amp-` prefix to prevent collision with other plugins.

File: tests/e2e/specs/admin/after-plugin-activation.js
Comment (westonruter) [inline tests/e2e/specs/admin/after-plugin-activation.js]: This should be updated to use the new bundled plugin, correct?

File: tests/e2e/specs/admin/after-plugin-activation.js
Comment (westonruter) [inline tests/e2e/specs/admin/after-plugin-activation.js]: The `visitAdminPage` function takes two arguments. Should an empty string be provided for the second?

File: assets/src/admin/site-scan-notice/notice.js
Comment (westonruter) [inline assets/src/admin/site-scan-notice/notice.js]: Let's instead say something like:

> AMP compatibility issues discovered with the following plugins:

Then let's move the “Review Plugin Suppression Settings” to be a button alongside the “View Compatible Plugins List” button. But we can shorten it to just “Review Plugin Suppression”.

(Otherwise, shouldn't `createInterpolateElement` be used? Or is that not safe since we may not be able to polyfill on this screen?)

File: assets/src/admin/site-scan-notice/notice.js
Comment (westonruter) [inline assets/src/admin/site-scan-notice/notice.js]:
```suggestion
						{ __( 'View AMP-Compatible Plugins', 'amp' ) }
```

File: assets/src/admin/site-scan-notice/notice.js
Comment (westonruter) [inline assets/src/admin/site-scan-notice/notice.js]:
```suggestion
				{ __( 'Checking your site for AMP compatibility issues…', 'amp' ) }
```

File: assets/src/admin/site-scan-notice/notice.js
Comment (westonruter) [inline assets/src/admin/site-scan-notice/notice.js]:
```suggestion
					{ __( 'No AMP compatibility issues detected.', 'amp' ) }
```

File: assets/src/admin/site-scan-notice/notice.js
Comment (westonruter) [inline assets/src/admin/site-scan-notice/notice.js]:
```suggestion
					{ __( 'AMP could not check your site for compatibility issues.', 'amp' ) }
```

File: assets/src/admin/site-scan-notice/notice.js
Comment (westonruter) [reply assets/src/admin/site-scan-notice/notice.js]: Here's what I'm suggesting:

image

File: assets/src/admin/site-scan-notice/index.js
Comment (westonruter) [inline assets/src/admin/site-scan-notice/index.js]:
```suggestion
				title={ __( 'The AMP Site Scanner has experienced an error.', 'amp' ) }
```

File: assets/src/admin/site-scan-notice/plugins-with-amp-incompatibility.js
Comment (westonruter) [inline assets/src/admin/site-scan-notice/plugins-with-amp-incompatibility.js]: See screenshot above for how I think this can be changed to have the plugin name at the beginning in bold, and then followed by “on X URLs”

File: assets/src/components/admin-notice/index.js
Comment (westonruter) [inline assets/src/components/admin-notice/index.js]: As noted in the stylesheet, these class names should all have an `amp-` prefix, right?

```suggestion
			className={ classnames( 'amp-admin-notice', className, {
				'amp-admin-notice--dismissible': isDismissible,
				'amp-admin-notice--info': type === ADMIN_NOTICE_TYPE_INFO,
				'amp-admin-notice--success': type === ADMIN_NOTICE_TYPE_SUCCESS,
				'amp-admin-notice--warning': type === ADMIN_NOTICE_TYPE_WARNING,
				'amp-admin-notice--error': type === ADMIN_NOTICE_TYPE_ERROR,
```

File: assets/src/components/admin-notice/index.js
Comment (westonruter) [inline assets/src/components/admin-notice/index.js]:
```suggestion
					className="amp-admin-notice__dismiss"
```

File: includes/validation/class-amp-validation-manager.php
Comment (westonruter) [inline includes/validation/class-amp-validation-manager.php:102]: I searched for the deleted code in this file on wpdirectory.net and didn't find any other plugins calling it. So safe to delete.

File: assets/src/components/admin-notice/index.js
Comment (westonruter) [reply assets/src/components/admin-notice/index.js]: Question: Any reason why core's class names couldn't be re-used? The core-provided activation notice is:

```html
<div id="message" class="updated notice is-dismissible">
    <p>Plugin activated.</p>
    <button type="button" class="notice-dismiss">
    <span class="screen-reader-text">Dismiss this notice.</span></button>
</div>
```

It seems like notice could use the same class names as `notice` and either `notice-info`, `notice-warning`, or `notice-error`?

File: tests/php/src/Admin/PluginActivationSiteScanTest.php
Comment (westonruter) [inline tests/php/src/Admin/PluginActivationSiteScanTest.php]:
```suggestion
			$this->markTestSkipped( 'WP install lacks dependency support.' );
```

**Review by westonruter — CHANGES_REQUESTED** [review]

File: assets/src/components/admin-notice/index.js
Comment (delawski) [reply assets/src/components/admin-notice/index.js]: The core notice component was my starting point. It turned out, however, that marrying the [existing jQuery-based logic](https://github.com/WordPress/WordPress/blob/master/wp-admin/js/common.js#L1085-L1116) with the React is quite cumbersome. Since the notice is pretty simple, I decided to recreate it as a React component.

File: tests/e2e/specs/admin/after-plugin-activation.js
Comment (delawski) [reply tests/e2e/specs/admin/after-plugin-activation.js]: Yes, I was just waiting for #6747 to get merged into `develop` so that the new demo plugin can be used here.

File: assets/src/components/admin-notice/style.scss
Comment (delawski) [reply assets/src/components/admin-notice/style.scss]: This has been addressed in 0f2c38a.

File: assets/src/components/admin-notice/index.js
Comment (delawski) [reply assets/src/components/admin-notice/index.js]: The component is now renamed to `AmpAdminNotice` (0f2c38a).

File: tests/e2e/specs/admin/after-plugin-activation.js
Comment (delawski) [reply tests/e2e/specs/admin/after-plugin-activation.js]: Yes, makes sense. In a5e649f I've updated this and other calls to `visitAdminPage` so that always two parameters are passed to it.

File: tests/e2e/specs/admin/after-plugin-activation.js
Comment (delawski) [reply tests/e2e/specs/admin/after-plugin-activation.js]: Updated in 0954bd4.

**Comment by delawski** [comment]
@westonruter I've updated the post-scan notice according to your suggestions. Also, the message for non-technical users has been updated.

**One plugin causing AMP incompatibilities**
| Technical user | Non-technical user |
| --- | --- |
| Screenshot 2021-12-01 at 12 40 51 | Screenshot 2021-12-01 at 12 48 24 |

**Two or more plugins causing AMP incompatibilities**
| Technical user | Non-technical user |
| --- | --- |
| Screenshot 2021-12-01 at 12 41 16 | Screenshot 2021-12-01 at 12 48 00 |

File: assets/src/admin/site-scan-notice/notice.js
Comment (delawski) [reply assets/src/admin/site-scan-notice/notice.js]: See my [latest comment](https://github.com/ampproject/amp-wp/pull/6685#issuecomment-983612091).

Comment (delawski) [reply assets/src/admin/site-scan-notice/notice.js]: > (Otherwise, shouldn't `createInterpolateElement` be used? Or is that not safe since we may not be able to polyfill on this screen?)

From what I can tell so far, the `createInterpolateElement` is useful when interpolating anything that is non-trivial. E.g. `<a href="#">...</a>` or `<b>...</b>` are simple elements and for them `dangerouslySetInnerHTML` + `sprintf` is sufficient. If we wanted to interpolate a React component or an HTML element with some conditional/complex attributes (like an external link), then `createInterpolateElement` would be the way to go for sure.

**Comment by milindmore22** [comment]
When I tried to activate any plugin, I got the following error
```json
{
  "message": "destroy is not a function",
  "stack": "TypeError: destroy is not a function\n    at commitHookEffectListUnmount (https://amp-local.test/wp-includes/js/dist/vendor/react-dom.js?ver=16.13.1:19845:13)\n    at commitPassiveHookEffects (https://amp-local.test/wp-includes/js/dist/vendor/react-dom.js?ver=16.13.1:19903:13)\n    at HTMLUnknownElement.callCallback (https://amp-local.test/wp-includes/js/dist/vendor/react-dom.js?ver=16.13.1:182:16)\n    at Object.invokeGuardedCallbackDev (https://amp-local.test/wp-includes/js/dist/vendor/react-dom.js?ver=16.13.1:231:18)\n    at invokeGuardedCallback (https://amp-local.test/wp-includes/js/dist/vendor/react-dom.js?ver=16.13.1:286:33)\n    at flushPassiveEffectsImpl (https://amp-local.test/wp-includes/js/dist/vendor/react-dom.js?ver=16.13.1:22988:11)\n    at unstable_runWithPriority (https://amp-local.test/wp-includes/js/dist/vendor/react.js?ver=16.13.1:2685:14)\n    at runWithPriority$1 (https://amp-local.test/wp-includes/js/dist/vendor/react-dom.js?ver=16.13.1:11174:12)\n    at flushPassiveEffects (https://amp-local.test/wp-includes/js/dist/vendor/react-dom.js?ver=16.13.1:22955:14)\n    at https://amp-local.test/wp-includes/js/dist/vendor/react-dom.js?ver=16.13.1:22834:13"
}
```
image

**Comment by delawski** [comment]
@milindmore22 Thank you for the report. I'm not able to reproduce the error. It happens only if Gutenberg is not activated. I'm investigating it now.

**Comment by delawski** [comment]
The issue described by @milindmore22 was caused by the fact that Gutenberg comes with React 17 while WordPress 5.8 is contains React 16.13. Version 17 brought [some changes](https://github.com/facebook/react/blob/main/CHANGELOG.md#1700-october-20-2020) to how `useEffect` cleanup functions are handled. With 18178b1, the Site Scan feature should work well with the older version of React, too. What's more, we'll now be testing if the feature works as expected even without Gutenberg (f2a0937).

Along the way, I noticed some other issues:
- `wp-components` has not been set as a dependency to the new `amp-site-scan-notice` stylesheet causing some minor visual issues (fixed in f6f5ff5).
- Site scan has been needlessly triggered on a plugin network activation (fixed in b17b93a).

**Comment by delawski** [comment]
While working on this issue, I noticed a potential bug with a multisite setup: #6755

**Review by milindmore22 — APPROVED** [review]
LGTM 👍🏼

File: assets/src/admin/site-scan-notice/notice.js
Comment (westonruter) [reply assets/src/admin/site-scan-notice/notice.js]: Isn't `createInterpolateElement` also preferable for security reasons since it avoids the use of `RawHTML` or `dangerouslySetInnerHTML`? This is what I understood from https://make.wordpress.org/core/2020/07/17/introducing-createinterpolateelement/

File: assets/src/components/site-scan-context-provider/index.js
Comment (westonruter) [inline assets/src/components/site-scan-context-provider/index.js:336]: I don't think this will ever be false because `window.clearTimeout` is a function.

Comment (westonruter) [reply assets/src/components/site-scan-context-provider/index.js:336]: Ah, nevermind. I suppose not since `clearTimeout` was defined above in the local scope.

File: assets/src/admin/site-scan-notice/plugins-with-amp-incompatibility.js
Comment (westonruter) [inline assets/src/admin/site-scan-notice/plugins-with-amp-incompatibility.js]: I appreciate the concern for users who are not technical, but the additional information  exposed below doesn't have anything technical on it. It just has the URLs on which issues were encountered. Listing out the URLs on which errors occur is helpful even for non-technical users so that they can check them out to see if anything is broken.  Therefore I don't think this condition is needed.

Something I can see as being helpful in terms of making it more user-friendly is for the original `<title>` of the validated URL to be stored and for that to be displayed as the link text instead of URL. But we can explore that in a subsequent improvement.

**Review by westonruter — APPROVED** [review]
This is so great and so much better than before. Not only does it no longer slow down the activation of a plugin, it also scans all the scannable URLs instead of just one. 💯

File: assets/src/admin/site-scan-notice/notice.js
Comment (delawski) [inline assets/src/admin/site-scan-notice/notice.js:96]: Note that [there's no need to explicitly add `noopener` if `noreferrer` is already there](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/noreferrer).

File: assets/src/admin/site-scan-notice/plugins-with-amp-incompatibility.js
Comment (delawski) [inline assets/src/admin/site-scan-notice/plugins-with-amp-incompatibility.js:74]: I guess it'd be best to use `ExternalLink` from `@wordpress/components` instead of a plain `<a>` tag.

File: assets/src/components/site-scan-context-provider/index.js
Comment (delawski) [reply assets/src/components/site-scan-context-provider/index.js:336]: You just made me realize that I did an embarrassing mistake here 🤦‍♂️

This code won't clear any timeout since it's not how timeouts are cleared in JS. The `setTimeout` returns an _identifier_ of a timeout and then you should pass that identifier to the `window.clearTimeout` function.

I'll fix it in a followup PR.

File: tests/e2e/specs/admin/after-plugin-activation.js
Comment (delawski) [inline tests/e2e/specs/admin/after-plugin-activation.js:24]: We can now safely remove this nested `describe` since we're not testing different user types anymore.

File: assets/src/admin/site-scan-notice/notice.js
Comment (delawski) [reply assets/src/admin/site-scan-notice/notice.js]: That's a very good point that I missed. Security-wise, it's better now to use `createInterpolateElement` instead of `dangerouslySetInnerHTML`. We had to resort to `dangerouslySetInnerHTML` before since there was no better option.

I think I'll issue a PR where the occurrences of `dangerouslySetInnerHTML` will be replaced with `createInterpolateElement` in the entire codebase.

File: assets/src/admin/site-scan-notice/notice.js
Comment (delawski) [reply assets/src/admin/site-scan-notice/notice.js:96]: That being said, it seems that `noopener noreferrer` is a common practice, e.g. the `<ExternalLink>` from `@wordpress/components` [sets both rel types](https://github.com/WordPress/gutenberg/pull/1577#pullrequestreview-50302342) (plus an `external` type).

If that's the case, let's keep both `noopener` and `noreferrer`.

File: assets/src/admin/site-scan-notice/plugins-with-amp-incompatibility.js
Comment (delawski) [reply assets/src/admin/site-scan-notice/plugins-with-amp-incompatibility.js:74]: Scratch that. `ExternalLink` adds an icon next to a link and there seem to be no way to remove the icon at the JS-level (other than hiding it with CSS):

Screenshot 2021-12-03 at 11 50 51

File: tests/e2e/specs/admin/after-plugin-activation.js
Comment (delawski) [reply tests/e2e/specs/admin/after-plugin-activation.js:24]: This has been addressed in a followup PR: #6758.

File: assets/src/components/site-scan-context-provider/index.js
Comment (delawski) [reply assets/src/components/site-scan-context-provider/index.js:336]: See: #6758.

File: assets/src/admin/site-scan-notice/notice.js
Comment (delawski) [reply assets/src/admin/site-scan-notice/notice.js]: I've created an issue for that: #6759.

**Pull request merged**
